pkg/controller/applicationlayer/applicationlayer_controller.go

```diff
@@ -0,0 +1,319 @@
+			return fmt.Errorf("applicationlayer-controller failed to watch ConfigMap %s: %v", configMapName, err)
+		}
+	}
+	// Watch for changes to FelixConfiguration.
+	err = c.Watch(&source.Kind{Type: &crdv1.FelixConfiguration{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return fmt.Errorf("applicationlayer-controller failed to watch FelixConfiguration resource: %w", err)
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileCompliance implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileApplicationLayer{}
+
+// ReconcileApplicationLayer reconciles a ApplicationLayer object
+type ReconcileApplicationLayer struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client          client.Client
+	scheme          *runtime.Scheme
+	provider        operatorv1.Provider
+	status          status.StatusManager
+	clusterDomain   string
+	licenseAPIReady *utils.ReadyFlag
+}
+
+// Reconcile reads that state of the cluster for a ApplicationLayer object and makes changes
+// based on the state read and what is in the ApplicationLayer.Spec
+func (r *ReconcileApplicationLayer) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ApplicationLayer")
+
+	applicationLayer, err := getApplicationLayer(ctx, r.client)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			reqLogger.Info("ApplicationLayer object not found")
+			r.status.OnCRNotFound()
+			return reconcile.Result{}, nil
+		}
+		r.status.SetDegraded("Error querying for Application Layer", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	preDefaultPatchFrom := client.MergeFrom(applicationLayer.DeepCopy())
+	// Set defaults
+	updateApplicationLayerWithDefaults(applicationLayer)
+
+	// Validate the configuration
+	if err := validateApplicationLayer(applicationLayer); err != nil {
+		r.status.SetDegraded("Invalid applicationLayer provided", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	// Write the application layer back to the datastore, so the controllers depending on this can reconcile.
+	if err = r.client.Patch(ctx, applicationLayer, preDefaultPatchFrom); err != nil {
+		log.Error(err, "Failed to write defaults")
+		r.status.SetDegraded("Failed to write defaults", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	variant, installation, err := utils.GetInstallation(ctx, r.client)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.status.SetDegraded("Installation not found", err.Error())
+			return reconcile.Result{}, err
```

Comment:
I know we do this everywhere else for the installation controller, but I'm wondering if we should stop returning an error here. Normally if the resource isn't found we set degraded and just return, and just rely on watching the missing object to trigger the reconcile when the object is available.

@tmjd what do you think, should we stop returning errors on this?

Reply:
Yeah I think that makes sense that we don't need to return an error.
@penkeysuresh if you're not aware when the return includes an error the request that triggered the reconcile will cause a re-queue of the request with an increasing backed off and the request is dropped eventually if it keeps erroring.
But like Brian said we should be able to depend on the watch on the Installation to trigger us.

So yeah lets not return err in this specific case.